spring-session/src/main/java/org/springframework/session/data/mongo/MongoSessionRepository.java

```diff
@@ -0,0 +1,160 @@
+
+	private static final String ID_FIELD_NAME = "_id";
+
+	private final MongoOperations mongoOperations;
+	private final ApplicationEventPublisher eventPublisher;
+	private final MongoSessionSerializer sessionSerializer;
+	private final Integer maxInactiveIntervalInSeconds;
+	private final String collectionName;
+
+	public MongoSessionRepository(MongoOperations mongoOperations, ApplicationEventPublisher eventPublisher,
+	                              MongoSessionSerializer mongoSessionSerializer, Integer maxInactiveIntervalInSeconds, String collectionName) {
+		this.mongoOperations = mongoOperations;
+		this.eventPublisher = eventPublisher;
+		this.sessionSerializer = mongoSessionSerializer;
+		this.maxInactiveIntervalInSeconds = maxInactiveIntervalInSeconds;
+		this.collectionName = collectionName;
+	}
+
+	@Override
+	public ExpiringSession createSession() {
+		MapSession session = new MapSession(UUID.randomUUID().toString().replaceAll("-", ""));
+		session.setMaxInactiveIntervalInSeconds(maxInactiveIntervalInSeconds);
+		return session;
+	}
+
+	@Override
+	public void save(ExpiringSession session) {
+		DBObject jo = sessionSerializer.serializeSession(session);
+		mongoOperations.getCollection(collectionName).save(jo);
+		eventPublisher.publishEvent(new SessionCreatedEvent(this, session.getId()));
+	}
+
+	@Override
+	public ExpiringSession getSession(String id) {
+		DBObject sessionWrapper = findSession(id);
+		if (sessionWrapper == null) {
+			return null;
+		}
+		ExpiringSession session = sessionSerializer.deserializeSession(sessionWrapper);
+		if (session.isExpired()) {
+			expireSession(id);
+			return null;
+		}
+		return session;
+	}
+
+	/**
+	 * Currently this repository allows only querying against {@code PRINCIPAL_NAME_INDEX_NAME}
+	 *
+	 * @param indexName  the name if the index (i.e. {@link FindByIndexNameSessionRepository#PRINCIPAL_NAME_INDEX_NAME})
+	 * @param indexValue the value of the index to search for.
+	 * @return sessions map
+	 */
+	@Override
+	public Map<String, ExpiringSession> findByIndexNameAndIndexValue(String indexName, String indexValue) {
+		if (!PRINCIPAL_NAME_INDEX_NAME.equals(indexName)) {
+			return Collections.emptyMap();
+		}
+		HashMap<String, ExpiringSession> result = new HashMap<String, ExpiringSession>();
+		List<DBObject> mapSessions = mongoOperations.find(getPrincipalQuery(indexValue), DBObject.class, collectionName);
+		for (DBObject session : mapSessions) {
+			ExpiringSession mapSession = sessionSerializer.deserializeSession(session);
+			result.put(mapSession.getId(), mapSession);
+		}
+		return result;
+	}
+
+	@Override
+	public void delete(String id) {
+		mongoOperations.remove(findSession(id), collectionName);
+		eventPublisher.publishEvent(new SessionDeletedEvent(this, id));
+	}
+
+	/**
+	 * Method cleanups expired sessions and emits {@code SessionExpiredEvent} event for expiring sessions
+	 */
+	@Scheduled(cron = "0 * * * * *")
```

Comment:
It may be good to have an expiration on the actual object as well to ensure that if the application is down when sessions expire (then brought back up) that the sessions are still cleaned up.